File: 11518 Dominos 2/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#define MAXN 10000

short**knock, knockCount[MAXN + 1];
bool visited[MAXN + 1];

void dfs(int start, int *count) {
    visited[start] = true;
    ++(*count);
    for(short i = 0; i < knockCount[start]; ++i) {
        if(!visited[knock[start][i]])
            dfs(knock[start][i], count);
    }
}

int main() {
    knock = (short**)malloc((MAXN + 1) * sizeof(short*));
    for(short i = 1; i <= MAXN; ++i)
        knock[i] = NULL;
    int cases;
    scanf("%d", &cases);
    while(cases--) {
        short n, m, l;
        scanf("%hd %hd %hd", &n, &m, &l);
        memset(knockCount, 0, sizeof(knockCount));
        while(m--) {
            short x, y;
            scanf("%hd %hd", &x, &y);
            knock[x] = (short*)realloc(knock[x], (knockCount[x] + 1) * sizeof(short));
            knock[x][knockCount[x]++] = y;
        }
        memset(visited, 0, sizeof(visited));
        int count = 0;
        while(l--) {
            short z;
            scanf("%hd", &z);
            if(!visited[z])
                dfs(z, &count);
        }
        printf("%d\n", count);
    }
    for(short i = 1; i <= MAXN; ++i)
        free(knock[i]);
    free(knock);
	return 0;
}

File: 11518 Dominos 2/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static int MAXN = 10000;
    static ArrayList<Short>[] knock;
    static boolean[] visited;
	public static void main(String[] args) throws IOException {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        st.nextToken();
        int cases = (int)st.nval;
        StringBuilder output = new StringBuilder();
        while(cases-- > 0) {
            st.nextToken();
            short n = (short)st.nval;
            st.nextToken();
            short m = (short)st.nval;
            st.nextToken();
            short l = (short)st.nval;
            knock = (ArrayList<Short>[])new ArrayList[n + 1];
            for(int i = 1; i <= n; ++i)
                knock[i] = new ArrayList<>();
            while(m-- > 0) {
                st.nextToken();
                short x = (short)st.nval;
                st.nextToken();
                short y = (short)st.nval;
                knock[x].add(y);
            }
            visited = new boolean[n + 1];
            int[] count = new int[1];
            while(l-- > 0) {
                st.nextToken();
                short z = (short)st.nval;
                if(!visited[z])
                    dfs(z, count);
            }
            output.append(count[0]).append('\n');
    	}
        System.out.print(output);
	}
	
	static void dfs(short start, int[] count) {
	    visited[start] = true;
	    ++count[0];
	    for(Short s : knock[start]) {
	        if(!visited[s])
	            dfs(s, count);
	    }
	}
}
